File: e2e/harness/harness.go
//go:build e2e

package harness

import (
	"context"
	"fmt"
	"net"
	"net/http"
	"net/http/cookiejar"
	"net/url"
	"os"
	"os/exec"
	"path/filepath"
	"syscall"
	"testing"
	"time"
)

// readinessTimeout bounds how long a spawned server may take to answer its
// first discovery request.
const readinessTimeout = 15 * time.Second

// shutdownTimeout bounds how long a stopped server may take to exit after
// its termination signal before it is killed.
const shutdownTimeout = 10 * time.Second

// Harness drives one isolated black-box Zen IdP instance for a single test:
// a dedicated configuration and state database in a per-test directory, one
// loopback port, and one spawned server process.
type Harness struct {
	dir     string
	baseURL string
	cmd     *exec.Cmd
	logFile *os.File
	browser *Browser
	cfg     Config
}

// New starts a fresh Zen IdP instance for one test: it validates and writes
// the YAML configuration of cfg into a per-test directory, binds a free
// loopback port, spawns the compiled executable with the matching
// environment, and waits until the discovery endpoint answers. It registers
// the teardown that stops the server when the test finishes.
func New(t *testing.T, cfg Config) *Harness {
	t.Helper()
	cfg.validate(t)

	dir := t.TempDir()
	port := freePort(t)
	issuer := fmt.Sprintf("http://127.0.0.1:%d", port)
	cfg.WriteFile(t, dir, issuer, port)

	jar, err := cookiejar.New(nil)
	if err != nil {
		t.Fatal(err)
	}
	origin, err := url.Parse(issuer)
	if err != nil {
		t.Fatal(err)
	}
	harness := &Harness{
		dir:     dir,
		baseURL: issuer,
		cfg:     cfg,
	}
	harness.browser = &Browser{
		http: &http.Client{
			Jar: jar,
			CheckRedirect: func(*http.Request, []*http.Request) error {
				return http.ErrUseLastResponse
			},
		},
		jar:    jar,
		origin: origin,
	}
	harness.start(t)
	t.Cleanup(func() { harness.stop(t) })
	return harness
}

// Browser returns the HTTP client of this harness. Its cookie jar is
// created once and shared across restarts, so browser state observed before
// a restart survives it.
func (h *Harness) Browser() *Browser {
	return h.browser
}

// BaseURL returns the issuer origin of this harness.
func (h *Harness) BaseURL() string {
	return h.baseURL
}

// Restart stops the server and starts it again with the same configuration
// and state database, simulating an ordinary service restart.
func (h *Harness) Restart(t *testing.T) {
	t.Helper()
	h.stop(t)
	h.start(t)
}

// start spawns the server with this harness's configuration and waits until
// it is ready, failing the test when the server cannot start or does not
// become ready in time.
func (h *Harness) start(t *testing.T) {
	t.Helper()
	logFile, err := os.Create(filepath.Join(h.dir, "server.log"))
	if err != nil {
		t.Fatal(err)
	}
	cmd := exec.CommandContext(t.Context(), BinaryPath(t), "serve")
	cmd.Dir = h.dir
	cmd.Env = append(os.Environ(),
		"ZEN_IDP_CONFIG_PATH="+filepath.Join(h.dir, "config.yaml"),
		"ZEN_IDP_SECRET="+h.cfg.RootSecret,
		"ZEN_IDP_DB_PATH="+filepath.Join(h.dir, "state.db"),
	)
	cmd.Stdout = logFile
	cmd.Stderr = logFile
	if err := cmd.Start(); err != nil {
		_ = logFile.Close()
		t.Fatal(err)
	}
	h.cmd = cmd
	h.logFile = logFile
	if !waitReady(t, h.baseURL) {
		t.Fatalf("server did not become ready:\n%s", h.serverLog())
	}
}

// stop terminates the server gracefully and waits for it to exit, killing
// it when it does not stop in time. It reports the server's own output when
// the surrounding test failed, so failures can be diagnosed from the
// process side.
func (h *Harness) stop(t *testing.T) {
	t.Helper()
	if h.cmd == nil {
		return
	}
	_ = h.cmd.Process.Signal(syscall.SIGTERM)
	done := make(chan error, 1)
	go func() { done <- h.cmd.Wait() }()
	select {
	case <-done:
	case <-time.After(shutdownTimeout):
		_ = h.cmd.Process.Kill()
		<-done
	}
	h.cmd = nil
	if t.Failed() {
		t.Logf("server output:\n%s", h.serverLog())
	}
	_ = h.logFile.Close()
	h.logFile = nil
}

// serverLog returns the captured output of the server process.
func (h *Harness) serverLog() string {
	contents, err := os.ReadFile(filepath.Join(h.dir, "server.log"))
	if err != nil {
		return fmt.Sprintf("(read server log: %v)", err)
	}
	return string(contents)
}

// freePort returns a free loopback TCP port for the server to bind.
func freePort(t *testing.T) int {
	t.Helper()
	var listenConfig net.ListenConfig
	listener, err := listenConfig.Listen(context.Background(), "tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	port := listener.Addr().(*net.TCPAddr).Port
	if err := listener.Close(); err != nil {
		t.Fatal(err)
	}
	return port
}

// waitReady polls the discovery endpoint until it answers with 200 or the
// readiness timeout elapses.
func waitReady(t *testing.T, baseURL string) bool {
	t.Helper()
	client := &http.Client{Timeout: time.Second}
	deadline := time.Now().Add(readinessTimeout)
	for time.Now().Before(deadline) {
		ready, err := probeReady(client, baseURL)
		if err == nil && ready {
			return true
		}
		time.Sleep(100 * time.Millisecond)
	}
	return false
}

// probeReady reports whether the discovery endpoint of the given origin
// answers with 200.
func probeReady(client *http.Client, baseURL string) (bool, error) {
	request, err := http.NewRequestWithContext(
		context.Background(),
		http.MethodGet,
		baseURL+"/.well-known/openid-configuration",
		nil,
	)
	if err != nil {
		return false, err
	}
	response, err := client.Do(request)
	if err != nil {
		return false, err
	}
	defer func() { _ = response.Body.Close() }()
	return response.StatusCode == http.StatusOK, nil
}

File: e2e/browser/harness/harness.ts
// Drives one isolated black-box Zen IdP instance for a single test: a
// dedicated configuration and state database in a per-test directory, one
// loopback port, and one spawned server process. Every instance is fully
// independent, so any number of tests can run in parallel without
// colliding, mirroring the e2e/http/harness Go package.

import { assertBinary, binaryPath } from "./binary.ts";
import { Client } from "./client.ts";
import { buildConfigDocument, type ConfigDocument, dumpYaml, validateConfig } from "./config.ts";

/** How long a spawned server may take to answer its first discovery
 * request before the test fails. */
const readinessTimeoutMs = 15_000;

/** How long a stopped server may take to exit after its termination
 * signal before it is killed. */
const shutdownTimeoutMs = 10_000;

/** How long to wait between readiness probes. */
const probeIntervalMs = 100;

/**
 * Root secret of instances that do not declare one. It satisfies the
 * service minimum length and stays stable, so tests can derive TOTP
 * secrets from it.
 */
export const defaultRootSecret = "0102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f20";

/** Options that describe one black-box instance. */
export interface HarnessOptions {
  /**
   * The YAML configuration document of the instance. The harness fills
   * the issuer and listener settings itself.
   */
  config: ConfigDocument;
  /** Root secret used as ZEN_IDP_SECRET; defaults to defaultRootSecret. */
  rootSecret?: string;
  /** Extra environment variables for the server process. */
  env?: Record<string, string>;
}

/**
 * Drives one isolated black-box Zen IdP instance for a single test: a
 * dedicated configuration and state database in a per-test directory, one
 * loopback port, and one spawned server process.
 */
export class Harness {
  /**
   * Starts a fresh instance for one test: allocates a per-test directory
   * and loopback port, writes the configuration, spawns the compiled
   * executable, and waits until the discovery endpoint answers. It stops
   * the instance and removes its directory when the process fails to
   * start, and the caller must eventually call stop.
   */
  static async start(options: HarnessOptions): Promise<Harness> {
    assertBinary();
    validateConfig(options.config);
    const dir = await Deno.makeTempDir({ prefix: "zen-idp-e2e-" });
    const port = freePort();
    const baseURL = `http://127.0.0.1:${port}`;
    const harness = new Harness(
      dir,
      baseURL,
      port,
      options.rootSecret ?? defaultRootSecret,
      options.env,
    );
    try {
      await harness.writeConfig(options.config);
      harness.startProcess();
      await harness.waitReady();
    } catch (error) {
      await harness.stop();
      throw error;
    }
    return harness;
  }

  private constructor(
    readonly dir: string,
    readonly baseURL: string,
    readonly port: number,
    readonly rootSecret: string,
    env: Record<string, string> | undefined,
  ) {
    this.#env = env;
    this.client = new Client(baseURL);
  }

  /**
   * HTTP client of this instance. Its cookie jar is created once and
   * shared across restarts, so browser state observed before a restart
   * survives it.
   */
  readonly client: Client;

  readonly #env: Record<string, string> | undefined;
  #process: Deno.ChildProcess | null = null;
  #logChunks: Uint8Array[] = [];

  /**
   * Stops the server and removes the instance directory. Safe to call
   * more than once.
   */
  async stop(): Promise<void> {
    await this.stopProcess();
    await Deno.remove(this.dir, { recursive: true }).catch(() => {});
  }

  /**
   * Stops the server and starts it again with the same configuration and
   * state database, simulating an ordinary service restart.
   */
  async restart(): Promise<void> {
    await this.stopProcess();
    this.startProcess();
    await this.waitReady();
  }

  /**
   * Stops the server, replaces the instance configuration, and starts
   * again with the same state database and listener port, simulating a
   * configuration change followed by a restart.
   */
  async reconfigure(config: ConfigDocument): Promise<void> {
    validateConfig(config);
    await this.stopProcess();
    await this.writeConfig(config);
    this.startProcess();
    await this.waitReady();
  }

  /**
   * Stops the server, removes the state database, and starts again with a
   * fresh one, simulating state-file loss followed by a restart.
   */
  async resetState(): Promise<void> {
    await this.stopProcess();
    for (const name of ["state.db", "state.db-wal", "state.db-shm"]) {
      await Deno.remove(`${this.dir}/${name}`).catch(() => {});
    }
    this.startProcess();
    await this.waitReady();
  }

  /** Returns the captured output of the server process. */
  serverLog(): string {
    return new TextDecoder().decode(concatChunks(this.#logChunks));
  }

  /** Returns the YAML configuration of the instance. */
  async configYaml(): Promise<string> {
    try {
      return await Deno.readTextFile(`${this.dir}/config.yaml`);
    } catch {
      return "(configuration unavailable)";
    }
  }

  /** Writes the instance configuration with the harness-owned settings. */
  private async writeConfig(config: ConfigDocument): Promise<void> {
    const document = buildConfigDocument(config, this.baseURL, this.port);
    await Deno.writeTextFile(`${this.dir}/config.yaml`, dumpYaml(document));
  }

  /** Spawns the server process with this instance's configuration. */
  private startProcess(): void {
    this.#logChunks = [];
    const command = new Deno.Command(binaryPath, {
      args: ["serve"],
      cwd: this.dir,
      env: {
        ZEN_IDP_CONFIG_PATH: `${this.dir}/config.yaml`,
        ZEN_IDP_SECRET: this.rootSecret,
        ZEN_IDP_DB_PATH: `${this.dir}/state.db`,
        ...this.#env,
      },
      stdout: "piped",
      stderr: "piped",
    });
    const process = command.spawn();
    this.#process = process;
    // Both streams share one log: each chunk is appended as it arrives,
    // and failures after the process exits are ignored.
    for (const stream of [process.stdout, process.stderr]) {
      stream.pipeTo(
        new WritableStream<Uint8Array>({
          write: (chunk) => {
            this.#logChunks.push(chunk);
          },
        }),
      ).catch(() => {});
    }
  }

  /**
   * Terminates the server gracefully and waits for it to exit, killing it
   * when it does not stop in time.
   */
  private async stopProcess(): Promise<void> {
    if (this.#process !== null) {
      const process = this.#process;
      this.#process = null;
      try {
        process.kill("SIGTERM");
      } catch {
        // The process already exited.
      }
      const exited = await Promise.race([
        process.status.then(() => true),
        delay(shutdownTimeoutMs).then(() => false),
      ]);
      if (!exited) {
        try {
          process.kill("SIGKILL");
        } catch {
          // The process already exited.
        }
        await process.status;
      }
    }
  }

  /**
   * Polls the discovery endpoint until it answers with 200, failing fast
   * with the server output when the process dies before becoming ready.
   */
  private async waitReady(): Promise<void> {
    const deadline = Date.now() + readinessTimeoutMs;
    while (Date.now() < deadline) {
      try {
        const response = await fetch(
          `${this.baseURL}/.well-known/openid-configuration`,
          { signal: AbortSignal.timeout(1_000) },
        );
        if (response.status === 200) {
          await response.arrayBuffer();
          return;
        }
      } catch {
        // Not ready yet.
      }
      if (await this.processExited()) {
        throw new Error(`server exited before becoming ready:\n${this.serverLog()}`);
      }
    }
    throw new Error(`server did not become ready:\n${this.serverLog()}`);
  }

  /** Reports whether the server process has already exited. */
  private async processExited(): Promise<boolean> {
    const process = this.#process;
    if (process === null) {
      return false;
    }
    return await Promise.race([
      process.status.then(() => true),
      delay(probeIntervalMs).then(() => false),
    ]);
  }
}

/** Returns a free loopback TCP port for the server to bind. */
function freePort(): number {
  const listener = Deno.listen({ hostname: "127.0.0.1", port: 0 });
  const port = (listener.addr as Deno.NetAddr).port;
  listener.close();
  return port;
}

/** Concatenates byte chunks into one byte array. */
function concatChunks(chunks: Uint8Array[]): Uint8Array {
  const total = chunks.reduce((sum, chunk) => sum + chunk.length, 0);
  const result = new Uint8Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    result.set(chunk, offset);
    offset += chunk.length;
  }
  return result;
}

/** Resolves after the given number of milliseconds. */
function delay(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}
